conda_declarative/spec.py

```diff
@@ -0,0 +1,821 @@
+        """
+        return TomlSingleEnvironment.model_validate(
+            {
+                "about": {
+                    "name": env.name if env.name else "",
+                    "revision": "1",
+                    "description": "",
+                },
+                "config": {
+                    "channels": env.config.channels,
+                    "platforms": {env.platform: {}},
+                    "variables": env.variables,
+                },
+                "platform": env.platform,
+                "system_requirements": [],
+                "version": 1,
+                "dependencies": env.requested_packages,
+                "pypi_dependencies": env.external_packages.get("pip", []),
+            }
+        )
+
+    @classmethod
+    def from_dict(cls, obj: dict[str, Any]) -> TomlSpec:
+        """Ingest a dict (e.g. read from a TOML spec file) to build a TomlSpec.
+
+        Parameters
+        ----------
+        obj : dict[str, Any]
+            A dict which can be validated with TomlSingleEnvironment
+
+        Returns
+        -------
+        Environment
+            Environment object which can be used to update a conda environment
+        """
+        raise NotImplementedError
+
+    @property
+    def model(self) -> TomlSingleEnvironment:
+        """Generate a pydantic model from the TomlSpec.
+
+        Returns
+        -------
+        TomlSingleEnvironment
+            Model of the environment
+        """
+        if self._model is None or self._environment is None:
+            self._load()
+        return self._model
+
+    @property
+    def env(self) -> Environment:
+        """Generate an Environment from the TomlSpec.
+
+        Returns
+        -------
+        Environment
+            An Environment instance populated by the TomlSpec
+        """
+        if self._model is None or self._environment is None:
+            self._load()
+        return self._environment
+
+    @property
+    def environment(self) -> Environment:
+        """Alias for `self.env`.
+
+        Returns
+        -------
+        Environment
+            An Environment instance populated from the TOML file
+        """
+        return self.env
+
+
+class Author(BaseModel):
+    """A model which holds author information."""
+
+    name: str
+    email: EmailStr | None = None
+
+
+class About(BaseModel):
+    """A model which stores metadata about an environment.
+
+    `license` is an SPDX license expression: https://spdx.dev/learn/handling-license-info/
+    `license_files` is a PEP639-compliant expression: https://peps.python.org/pep-0639/#term-license-expression
+    """
+
+    model_config = ConfigDict(
+        alias_generator=lambda name: name.replace("_", "-"),
+        validate_by_name=True,
+        validate_by_alias=True,
+    )
+
+    name: str
+    revision: str
+    description: str
+    authors: list[Author] = []
+    license: str = ""
+    license_files: list[str] = []
```

Comment:
Down the line we can think about doing some validation here to ensure these are valid expressions.